### ISSUE TITLE
Undocumented Behavior of is_decreasing, is_increasing, is_strictly_decreasing, is_strictly_increasing, and is_subset

### ISSUE THREAD

**Issue by nlchar** [sympy/sympy#20745]
Hello.
I have noticed some undocumented behavior of these functions:

1. sympy.calculus.singularities.is_decreasing ([link to docs](https://docs.sympy.org/latest/modules/calculus/index.html#sympy.calculus.singularities.is_decreasing))
2. sympy.calculus.singularities.is_increasing ([link to docs](https://docs.sympy.org/latest/modules/calculus/index.html#sympy.calculus.singularities.is_increasing)) 
3. sympy.calculus.singularities.is_strictly_decreasing ([link to docs](https://docs.sympy.org/latest/modules/calculus/index.html#sympy.calculus.singularities.is_strictly_decreasing))
4. sympy.calculus.singularities.is_strictly_increasing ([link to docs](https://docs.sympy.org/latest/modules/calculus/index.html#sympy.calculus.singularities.is_strictly_increasing))
5. sets.Set.is_subset ([link to docs](https://docs.sympy.org/latest/modules/sets.html#sympy.sets.sets.Set.is_subset))

All these methods declared to return either `True` or `False` as their result. However, these return `None` in some circumstances. 
My example: 
```python
x = sympy.Symbol('x')
f = sympy.exp(sympy.sin(x))
intervals = [Interval.open(0, Pi/2), Interval.open(Pi/2, 3*Pi/2), Interval.open(3*Pi/2, 2*Pi)]
for interval in intervals:
    display(interval)
    print(is_monotonic(f, interval=interval, symbol=x), 
            is_decreasing(f, interval=interval, symbol=x), is_increasing(f, interval=interval, symbol=x),
            is_strictly_decreasing(f, interval=interval, symbol=x), is_strictly_increasing(f, interval=interval, symbol=x))
```
Result of the code execution are:
![image](https://user-images.githubusercontent.com/64870515/103652746-9d5f4800-4f74-11eb-9558-79df5ad82ea8.png)
The function (red) and its derivative (purple) with the points of interest ([link to desmos for interactive view](https://www.desmos.com/calculator/taegfbqemn)):
![image](https://user-images.githubusercontent.com/64870515/103652870-c7186f00-4f74-11eb-867c-66fb28b206f1.png)

The source of the functions can be found [here](https://github.com/sympy/sympy/blob/master/sympy/calculus/singularities.py).
Let me reproduce the actual code that runs (for instance, in case of the interval (0, Pi/2) and when checking for increase):
```python
x = sympy.Symbol('x')
f = sympy.exp(sympy.sin(x))
df = f.diff()
sol = sympy.solveset((lambda x: x>=0)(df), x, S.Reals)
print(Interval.open(0, Pi/2).is_subset(sol))
``` 
This code prints None, so does the sympy.calculus.singularities.is_increasing. 
The reasons is `is_subset`. Maybe it behaves in such a way because `sol` is a ConditionSet: 
![image](https://user-images.githubusercontent.com/64870515/103653512-ae5c8900-4f75-11eb-8a86-fc9f76e68613.png)

I do not know if it is a bug or just undocumented behavior, but in my opinion functions should not return `None`. If the software cannot tell you the answer, it'd better raise an exception, not just return `None`.

**Comment by nlchar** [comment]
My version is sympy==1.7.1

**Comment by Shardul555** [comment]
I think it is a bug as ```is_increasing``` function should have given true for interval(0,pi/2)

**Comment by oscarbenjamin** [comment]
There are lots of possibilities for improvement in either `is_increasing` or `is_subset` etc.

There will always be a need to return something in the indeterminate case and I think that the combination of True, False and None seems like a reasonable set of return values. That's also what happens in other sympy functions that give a bool but sometimes need to be indeterminate like `is_real` etc so it is actually a pattern across sympy that `None` means "unknown".

I'm not sure if that was the explicit intention for `is_increasing` though but for the sakes of argument let's presume that it was. It's important to be clear that returning `None` is not ideal but it is not a bug: returning True/False when the answer might not be True/False is a bug. Returning `None` is always logically consistent when `None` refers to the indeterminate case. It does however mean that there is room for improvement because it should be possible to return True/False in these particular examples.

**Comment by nlchar** [comment]
> There are lots of possibilities for improvement in either `is_increasing` or `is_subset` etc.
> 
> There will always be a need to return something in the indeterminate case and I think that the combination of True, False and None seems like a reasonable set of return values. That's also what happens in other sympy functions that give a bool but sometimes need to be indeterminate like `is_real` etc so it is actually a pattern across sympy that `None` means "unknown".
> 
> I'm not sure if that was the explicit intention for `is_increasing` though but for the sakes of argument let's presume that it was. It's important to be clear that returning `None` is not ideal but it is not a bug: returning True/False when the answer might not be True/False is a bug. Returning `None` is always logically consistent when `None` refers to the indeterminate case. It does however mean that there is room for improvement because it should be possible to return True/False in these particular examples.

Agree.
Maybe the implementation of is_increasing and others should be changed so that these cover such cases. 
For example, the implementation might rely on explicit calculations, like:
1. Find points such that df=0 and then find their interference with domain of interest (like R or some Set)
2. Check if df sign have no change on the domain.
3. Give result based on sign of df (if df have no sign change and sign = +1, then it increases; if sign changes or is -1 then return false.

So, in other words, can we change the implementation to avoid using is_subset? I am asking because I don't know how the symbolic compilations should be implemented properly.

**Comment by almighty123789** [comment]
`Pi=pi`
`x = sympy.Symbol('x')`
`f = sin(x)`
`intervals = [Interval.open(0, Pi/2), Interval.open(Pi/2, 3*Pi/2), Interval.open(3*Pi/2, 2*Pi)]`
`for interval in intervals:`
    `display(interval)`
    `print(is_monotonic(f, interval=interval, symbol=x),` 
            `is_decreasing(f, interval=interval, symbol=x), is_increasing(f, interval=interval, symbol=x),`
           ` is_strictly_decreasing(f, interval=interval, symbol=x), is_strictly_increasing(f, interval=interval, symbol=x))`
`sympy.__version__`

**Weirdly this works well with  "normal" sin(x) ( I haven't checked if the answers are correct, but atleast the answers are in Boolearn)**

![Screenshot (75)](https://user-images.githubusercontent.com/52477495/103906139-87dd5000-5125-11eb-85f4-2185527b6d51.png)

**Comment by almighty123789** [comment]
@oscarbenjamin  I checked and found  There is no is_subset implementation which takes in Interval and a ConditionSet as arguments (I am assuming this is becase sympy doesn't have a way to actually solve  a conditionset?) as to why it 'returns None' I found that the function does not return anything explicitly for these cases ( since all returns are in if blocks), it 's just how python returns  None when we call a function which does not return anything.

Is there a computational method for checking if the function  is decreasing or increasing ?

**Comment by oscarbenjamin** [comment]
> all returns are in if blocks

This is a common pattern in sympy because there are many functions that use None to mean unknown. The if blocks are used to check for cases where the answer is knowable. If none of those cases comes up then None can be returned implicitly.

**Comment by nlchar** [comment]
@oscarbenjamin The question is: is it possible to implement the methods with no base on is_subset?
For instance, does my above-mentioned solution apply?

**Comment by nlchar** [comment]
I mean that the current implementation is too abstract and high level and maybe we should consider making it more low-level and 'computationaly rich'.
But how can we do it? That is the question

**Comment by oscarbenjamin** [comment]
We could improve `solveset` for inequalities so that it does not give a `ConditionSet` for this case:
```
In [5]: solveset(exp(sin(x)).diff(x) > 0, x, Reals)  # Should be the same as Out[7]
Out[5]: 
⎧            ⎛ sin(x)           ⎞⎫
⎨x | x ∊ ℝ ∧ ⎝ℯ      ⋅cos(x) > 0⎠⎬
⎩                                ⎭

In [6]: solveset(exp(sin(x)) > 0, x, Reals)
Out[6]: ℝ

In [7]: solveset(cos(x) > 0, x, Reals)  # should be an `ImageSet`.
Out[7]: 
⎡   π⎞   ⎛3⋅π     ⎞
⎢0, ─⎟ ∪ ⎜───, 2⋅π⎟
⎣   2⎠   ⎝ 2      ⎠
```

**Comment by almighty123789** [comment]
@oscarbenjamin , One thing I have noticed is that the functioni is_monotonic works quite well and  is less prone to fail.
So, why not use this approach when checking if the function is increasing or decreasing. 
For example, in the OP, let's consider the interval ` pi/2 , 3pi/2`.  Now to check if the function is increasing or decreasing in the the specified interval, **We should first check if the function is even monotonic or not** ,since if it is not, then clearly, it cannot be increasing or decreasing within that inteval.

This is the  modified code of is_increasing,  and   is a crude implementation of what i wrote above.

```
if is_monotonic(expression,interval,symbol) is True:
        args=interval.args
        left=args[0]
        right=args[1]
        mid=(left+right)/2
        derivative=expression.diff(symbol)
        if derivative.subs(symbol,mid)>=0:
            return True
        else:
           return False
 else:
        return False
```
`>>> from sympy import *`
`>>> from sympy.abc import x,y`
`>>> f=exp(sin(x))`
`>>> Pi=pi`
`>>> intervals = [Interval.open(0, Pi/2), Interval.open(Pi/2, 3*Pi/2), Interval.open(3*Pi/2, 2*Pi)]`
`>>> for inter in intervals:`
`     ...  print(is_decreasing(f,inter,x),is_increasing(f,inter,x))`
`False True`
`True False`
`False True`


What are your thoughts?

**Comment by almighty123789** [comment]
@oscarbenjamin  any ideas as to how to go about to make these changes? This is definitely a problem of _solveset , 
the _solveset method calls solve_univariate inequality  which in turns calls function_range to find  which again calls solveset to find the critical points, at which point the function ends up throwing Not Implemented Error and solveset ends up returning the default conidtion set. 

This is because of this
````
elif f.is_Add:
        a, h = f.as_independent(symbol)
        m, h = h.as_independent(symbol, as_Add=False)
        if m not in {S.ComplexInfinity, S.Zero, S.Infinity,
                              S.NegativeInfinity}:
            f = a/m + h
````
This **does** not factor expressions which should have been factored,  for example  expressions like these
`` exp(sin(x))*cos(x) + exp(sin(x))*x = exp(sin(x))*(cos(x)+x) ``

A possible solution would have been to use 
`` f = f.together() ``, but  this is erronous, since the expression can be a rational function , or the common term might be a number. I tried to use this ,but this leads to several errors.

If such Add instances are able to converted into Mul then  the below of piece of code would successfully be able to give correct answers . 
```
elif f.is_Mul and all(_is_finite_with_finite_vars(m, domain)
            for m in f.args ):
         result = Union(*[solver(m, symbol) for m in f.args])
````
